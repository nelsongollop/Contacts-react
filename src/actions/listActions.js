import * as API from '../api/requests'

export function getContacts () {
    return{
        type: "GET_CONTACTS",
        payload: new Promise((resolve, reject) => {
            API.getContacts()
                .then(
                    (result) => {
                        resolve(result.data)
                    },
                    (error) => {
                        reject(error)
                    }
                )
        })
    }
}

export function deleteContact (id) {
    return{
        type: "GET_CONTACTS",
        payload: new Promise((resolve, reject) => {
            API.deleteContact(id)
                .then(
                    (result) => {
                        resolve(result.data)
                    },
                    (error) => {
                        reject(error)
                    }
                )
        })
    }
}

export function filter(keywords) {
    return {
        type: "FILTER_LIST",
        payload: keywords
    }
}

export function setUpdate(update){
    return{
        type: "SET_UPDATE",
        payload: update
    }
}

export function setSelected(index){
    return{
        type: "SET_SELECTED",
        payload: index
    }
}

export function setContacts(list){
    return{
        type: "SET_CONTACTS",
        payload: list
    }
}
