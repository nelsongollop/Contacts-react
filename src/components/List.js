import React, {useState, useEffect} from "react";
import { FixedSizeList as List } from "react-window";
import AutoSizer from 'react-virtualized-auto-sizer'
import Avatar from '@material-ui/core/Avatar';
import axios from 'axios'
import {connect} from 'react-redux'
import {getContacts} from "../actions/listActions";

function ContactList (props) {
    const [selected, setSelected] = useState(-1);
    useEffect(() => {
        props.getContacts()
    }, [])

    const Row = ({ index, style }) => (
        <div className={selected == index ? "list-item selected" : "list-item"} style={style} onClick={() => {
            setSelected(index)
            props.onSelect(props.listState.filtered[index])
        }}>
            <div className="col-md-1">
                <Avatar>NG</Avatar>
            </div>
            <div className="col-md-11">
                <ul className="info-list">
                    <li><h5>{props.listState.filtered[index].first_name} {props.listState.filtered[index].last_name}</h5></li>
                    <li>{props.listState.filtered[index].phone}</li>
                    <li>{props.listState.filtered[index].email}</li>
                </ul>
            </div>
        </div>
    );

    return (
        <AutoSizer>
            {({ height, width }) => (
                <List
                    className="List"
                    height={height}
                    itemCount={props.listState ? props.listState.filtered.length : 0}
                    itemSize={75}
                    width={width}
                >
                    {Row}
                </List>
            )}
        </AutoSizer>
    );
}
const mapStateToProps = (state) => {
    return {
        listState: state.list
    }
}
const mapDispatchToProps = (dispatch) => {
    return {
        getContacts: () => {
            dispatch(getContacts())
        }
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(ContactList)