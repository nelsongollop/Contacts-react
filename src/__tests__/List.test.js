import React from 'react';
import {Provider} from 'react-redux'
import configureStore from 'redux-mock-store';
import {configure, shallow, mount} from 'enzyme'
import Adapter from 'enzyme-adapter-react-16';
import List from '../components/List'
import * as Actions from "../actions/listActions";

configure({ adapter: new Adapter() });
const mockStore = configureStore([]);

let mockList = [
    {id: '0', first_name: 'Daenerys', last_name: 'Targaryen', phone: '123', email: 'a@a.com'},
    {id: '1', first_name: 'Jhon', last_name: 'Snow', phone: '123', email: 'a@a.com'}
]

describe('List', () => {
    let store;
    let wrapper;
    beforeEach(() => {
        store = mockStore();
        store.dispatch = jest.fn();
        jest.spyOn(Actions, 'getContacts').mockImplementation(() => Promise.resolve(mockList))

        wrapper = mount(
            <Provider store={store}>
                <List />
            </Provider>
        );
    });

    it('should render a List', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('Should call to get the list', () => {
        expect(store.dispatch).toHaveBeenCalledTimes(1);
        expect(store.dispatch).toHaveBeenCalledWith(
            Actions.getContacts()
        );
    });

    // it('Should populate the list', () => {
    //     expect(wrapper.find('.list-item')).toHaveLength(2)
    // });
    // it('Upon clicking a contact, it should show it', () => {
    //     wrapper.find('.list-item').simulate('click')
    //     expect(store.dispatch).toHaveBeenCalledTimes(2);
    //     expect(store.dispatch).toHaveBeenCalledWith(
    //         Actions.setUpdate(true),
    //         Actions.setSelected(0)
    //     );
    // });
});
