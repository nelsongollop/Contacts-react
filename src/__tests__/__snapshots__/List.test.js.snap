// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List should render a List 1`] = `ReactWrapper {}`;
